Enable reordering pass

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100746
* #100653
* #100652
* __->__ #100747

Restricts the reordering pass to only reorder nodes before the `copy_` epilogue that functionalization generates.

Brings `python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --amp --inference --only hf_Bert` from 1.46 -> 1.49



cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
Enable reordering pass

[ghstack-poisoned]

eellison committed:
Update on "Enable reordering pass"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Enable reordering pass"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Enable reordering pass"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Enable reordering pass"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Enable reordering pass"


Brings `python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --amp --inference --only hf_Bert` from 1.46 -> 1.49

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -81,8 +82,27 @@ def visit(other_node):
+
     for node in reversed(graph.nodes):
         seen_nodes.add(node)
+        if first_copy is not None:
```
im missing where 'first_copy' causes a change in reordering behavior.  it looks like you just find the first copy here and continue adding all the nodes to the graph?

eellison replied:
I'm only doing reordering before the copy_ epilogue (only reordering before the first copy_ in the graph)

wconstab commented:
I don't think this intersects with inductor collectives.  They are only mutating 'opportunistically' when the scheduler can safely reuse the input buffer to the collective.  The ops are 'functional' by schema, and don't result in a copy_ from functionalization.

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -81,8 +82,27 @@ def visit(other_node):
 
     seen_nodes = set()
+
+    # only reorder nodes before the first copy_ in the graph.
```
Since we iterate in reverse order, wouldn't this do after the first copy?  Which might include other copies...

eellison replied:
We are find the first copy not in reversed iteration, maybe last_copy would be clearer ?

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -51,6 +51,7 @@ def post_grad_passes(gm: torch.fx.GraphModule):
     if config.reordering:
         # has some issues with mutation in inference mode
         reorder_for_locality(gm.graph)
+        gm.recompile()
```
Should this go at the end?

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -81,8 +82,27 @@ def visit(other_node):
+        if first_copy is not None:
+            if first_copy is node:
+                first_copy = None
+            continue
```
Did you mean break here?  Won't this just skip the single node and continue?

eellison replied:
We don't do any reordering until we see the first copy

jansel replied:
I get it now, I think the original version is correct.

eellison committed:
Update on "Enable reordering pass"


Restricts the reordering pass to only reorder nodes before the `copy_` epilogue that functionalization generates.

Brings `python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --amp --inference --only hf_Bert` from 1.46 -> 1.49



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "braking internal builds" -c nosignal

eellison commented:
@pytorchbot merge
